ci: Ensure twister steps use the bsim version from the manifest

aescolar opened the pull request:
The west manifest may point to a newer version of bsim than the one avaliable in the docker image. Let's ensure we run with that one.
This is equivalent to the change done in 933d338c9783cce55466fed000baf2f3beeac891 for the bsim-tests workflow.
(The twister workflows are normally less sensitive to running on an older bsim version due to not having radio activity, but eventually things fail)

Note that the update and rebuild is incremental, so if the docker image already has the right version the operation is very fast (~1 second)

-----

Docker version update to the latest version: https://github.com/zephyrproject-rtos/docker-image/pull/178 (which after this PR can come calmly, and without this PR we are in a hurry to get into use )

aescolar committed:
ci: Ensure twister steps use the bsim version from the manifest

The west manifest may point to a newer version of bsim than
the one avaliable in the docker image. Let's ensure
we run with that one.
This is equivalent to the change done in
933d338c9783cce55466fed000baf2f3beeac891
for the bsim-tests workflow.

Note that the update and rebuild is incremental,
so if the docker image already has the right
version the operation is very fast
(~1 second)

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

fabiobaltieri commented on .github/workflows/twister.yaml:
```diff
@@ -213,6 +213,17 @@ jobs:
+          git fetch -n origin ${BSIM_VERSION}
+          git config --global advice.detachedHead false
+          git checkout ${BSIM_VERSION}
+          west update
```
hey I'm a bit confused, shouldn't the existin `west update` in `Environment Setup` step take care of this already?

aescolar replied:
For the Babblesim simulator we do not get it (and rebuild it) each time, but instead we keep a "preinstalled" version in the docker image in /opt. Apart from historic reasons, it saves a few minutes.
The simulator changes very rarely and an incremental fetch and build is ~1second vs a full build (several minutes).

fabiobaltieri replied:
oh because bsim is in the `babblesim` troup and the worfklow only does `ci` and `optional`? Ok but then is this `west update` command needed? I should not do anything since it's already been run before.

fabiobaltieri replied:
oookay.. I'll knock myself out :-)

aescolar replied:
It is a fair question :). Zephyr keeps nested in its manifest the bsim project manifest as a disabled group. Users are recommended to use that by just enabling that group.
But for CI we don't use it from there.
All we do here is check what version the Zephyr manifest is pointing at (store it in BSIM_VERSION); then change directory to /opt/bsim_west/bsim, where a standalone bsim can be found (without anything from Zephyr), and update that version to whatever Zephyr's manifest points to. (here `west` is used as a repository management tool, just like google's `repo` fully detached from Zephyr itself)
Normally BSIM_VERSION will just be the one prebuilt in the docker image, but sometimes (during updates) the manifest may be ahead of the docker image, or we may be running in a newer docker image with an older version of Zephyr for whatever reason.
With this minor trick, we both save time not fetching and rebuilding bsim in each CI runner, and ensure we run with the version pointed by the Zephyr manifest no matter the CI image.

fabiobaltieri replied:
Yeah got it now, I just got confused by that west command being there did not realize was another full checkout there, makes sense.

Then I think you could also pass `-o=--depth=1 -n` to make it download even less but it's likely milliseconds of difference in this case.

fabiobaltieri commented on .github/workflows/twister.yaml:
```diff
@@ -213,6 +213,17 @@ jobs:
+          echo "Manifest points to bsim sha $BSIM_VERSION"
+          cd /opt/bsim_west/bsim
+          git fetch -n origin ${BSIM_VERSION}
+          git config --global advice.detachedHead false
```
how about moving this with the other `git config` commands in the `Environment Setup` step? Just so everything git is configured at the start and this step does not have to mix config and normal operations

aescolar replied:
I just changed the command below to use that option for that command alone instead of setting it globally.
(to keep this localized to a single step, as it is replicated in several places and we may want to eventually refactor it out)

fabiobaltieri replied:
ok, though I think it would have been fine to change it for all commands, less spammy stuff in the log

aescolar committed:
ci: bsim update step: Do not affect global git config

When updating the bsim install in the docker image's
/opt/bsim_west to the version pointed in the Zephyr
manifest, we are detaching the head to that version.
We suppress the "You are in 'detached HEAD' state..."
git warning as it is just noise in CI.

So far we were doing it by globally supressing that
warning. Let's instead just supress it for the
command in question so we don't change global
git config against expectations.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>
